Gems/Atom/Utils/Code/Include/Atom/Utils/StableDynamicStructOfArrays.h

```diff
@@ -0,0 +1,389 @@
+
+        /// Used as an optimization to skip pages that are known to already be full. Generally this will point a page that has space available
+        /// in it, but it could point to a full page as long as there are no other available pages before that full page. When there are no
+        /// pages at all, this will point to nullptr. When all pages are full, this may point to any page, including the last page.
+        Page* m_firstAvailablePage = nullptr;
+
+        size_t m_pageCounter = 0; ///< The total number of pages that have been created (not how many currently exist).
+        size_t m_itemCount = 0; ///< The total number of items in this container.
+    };
+
+    /// Private class used by StableDynamicStructOfArrays to manage the arrays of data.
+    template<size_t ElementsPerPage, class Allocator, typename... value_types>
+    struct StableDynamicStructOfArrays<ElementsPerPage, Allocator, value_types...>::Page
+    {
+        static constexpr size_t InvalidPage = AZStd::numeric_limits<size_t>::max();
+        static constexpr uint64_t FullBits = 0xFFFFFFFFFFFFFFFFull;
+        static constexpr size_t NumUint64_t = ElementsPerPage / 64;
+        using ItemTupleType = AZStd::tuple<value_types *...>;
+        Page();
+        ~Page() = default;
+
+        /// Reserve the next availble index and return it. If no more space is available, returns InvalidPage.
+        size_t Reserve();
+
+        /// Free the given index so it can be reserved again.
+        void Free(ItemTupleType items);
+
+        /// True if this page is completely full
+        bool IsFull() const;
+
+        /// True if this page is completely empty
+        bool IsEmpty() const;
+
+        ItemTupleType GetItems(size_t index);
+
+        /**
+        * Gets a specific item from a specific row of the Page
+        * Note: may return empty slots.
+        */
+        template<size_t RowIndex>
+        auto* GetItem(size_t index);
+
+        /// Gets the number of items allocated on this page.
+        size_t GetItemCount() const;
+
+        size_t m_bitStartIndex = 0; ///< Index of the first uint64_t that might have space.
+        Page* m_nextPage = nullptr; ///< pointer to the next page.
+        StableDynamicStructOfArrays<ElementsPerPage, Allocator, value_types...>*
+            m_container; ///< pointer to the container this page was allocated from.
+        size_t m_pageIndex = 0; ///< used for comparing pages when items are freed so the earlier page in the list can be cached.
+        size_t m_itemCount = 0; ///< the number of items in the page.
+        AZStd::array<uint64_t, NumUint64_t> m_bits; ///< Bits representing free slots in the array. Free slots are 1, occupied slots are 0.
+
+        using PageTupleType = AZStd::tuple<AZStd::aligned_storage_t<ElementsPerPage * sizeof(value_types), alignof(value_types)>...>;
+        PageTupleType m_data; ///< aligned storage for all the actual data.
```

Comment:
[aligned_storage_t](https://stackoverflow.com/a/71828512) has been deprecated and instead we should something like `alignas(value_types) AZStd::byte[sizeof(value_types) * ElementsPerPage]`
```suggestion
        using PageTupleType = AZStd::tuple<AZStd::byte[ElementsPerPage * sizeof(value_types)] ...>;
        alignas(value_types...) PageTupleType m_data; ///< aligned storage for all the actual data.
```